brokers/vscode/broker.go

```diff
@@ -0,0 +1,296 @@
+}
+
+// NewBroker creates Che VS Code extension broker instance
+func NewBroker() *VSCodeExtensionBroker {
+	return &VSCodeExtensionBroker{
+		common.NewBroker(),
+		files.New(),
+		storage.New(),
+	}
+}
+
+// Start executes plugins metas processing and sends data to Che master
+func (broker *VSCodeExtensionBroker) Start(metas []model.PluginMeta) {
+	if ok, status := broker.Storage.SetStatus(model.StatusStarted); !ok {
+		m := fmt.Sprintf("Starting broker in state '%s' is not allowed", status)
+		broker.PubFailed(m)
+		broker.PrintFatal(m)
+	}
+	broker.PubStarted()
+	broker.PrintInfo("Started Plugin Broker")
+
+	broker.PrintPlan(metas)
+
+	broker.PrintInfo("Starting plugins processing")
+	for _, meta := range metas {
+		err := broker.processPlugin(meta)
+		if err != nil {
+			broker.PubFailed(err.Error())
+			broker.PrintFatal(err.Error())
+		}
+	}
+
+	if ok, status := broker.Storage.SetStatus(model.StatusDone); !ok {
+		err := fmt.Sprintf("Setting '%s' broker status failed. Broker has '%s' state", model.StatusDone, status)
+		broker.PubFailed(err)
+		broker.PrintFatal(err)
+	}
+
+	plugins, err := broker.Storage.Plugins()
+	if err != nil {
+		broker.PubFailed(err.Error())
+		broker.PrintFatal(err.Error())
+	}
+	pluginsBytes, err := json.Marshal(plugins)
+	if err != nil {
+		broker.PubFailed(err.Error())
+		broker.PrintFatal(err.Error())
+	}
+
+	broker.PrintInfo("All plugins have been successfully processed")
+	broker.PubDone(string(pluginsBytes))
+	broker.CloseConsumers()
+}
+
+// PushEvents sets given tunnel as consumer of broker events.
+func (broker *VSCodeExtensionBroker) PushEvents(tun *jsonrpc.Tunnel) {
+	broker.Broker.PushEvents(tun, model.BrokerStatusEventType, model.BrokerResultEventType, model.BrokerLogEventType)
+}
+
+func (broker *VSCodeExtensionBroker) processPlugin(meta model.PluginMeta) error {
+	broker.PrintDebug("Stared processing plugin '%s:%s'", meta.ID, meta.Version)
+	if meta.Attributes == nil || meta.Attributes["extension"] == "" {
+		m := fmt.Sprintf("VS Code extension field 'extension' is missing in description of plugin %s:%s", meta.ID, meta.Version)
+		return errors.New(m)
+	}
+	url := meta.Attributes["extension"]
+	image := meta.Attributes["container-image"]
+	if image == "" {
+		m := fmt.Sprintf("VS Code extension field 'container-image' is missing in description of plugin %s:%s", meta.ID, meta.Version)
+		return errors.New(m)
+	}
+
+	workDir, err := broker.ioUtil.TempDir("", "vscode-extension-broker")
+	if err != nil {
+		return err
+	}
+
+	archivePath := filepath.Join(workDir, "pluginArchive")
+	unpackedPath := filepath.Join(workDir, "plugin")
+
+	// Download an archive
+	broker.PrintDebug("Downloading archive '%s' for plugin '%s:%s' to '%s'", url, meta.ID, meta.Version, archivePath)
+	err = broker.download(url, archivePath, meta)
+	if err != nil {
+		return err
+	}
+
+	// Unzip it
+	broker.PrintDebug("Unzipping archive '%s' for plugin '%s:%s' to '%s'", url, meta.ID, meta.Version, unpackedPath)
```

Comment:
Should `pluginArchive` be first argument instead of `url`?

Reply:
It was supposed to be URL